src/solvers/refinement/string_refinement.cpp

```diff
@@ -1212,36 +1222,85 @@ void debug_model(
   stream << messaget::eom;
 }
 
+sparse_arrayt::sparse_arrayt(const with_exprt &expr)
+{
+  auto ref = std::ref(static_cast<const exprt &>(expr));
+  while(can_cast_expr<with_exprt>(ref.get()))
+  {
+    const auto &with_expr = expr_dynamic_cast<with_exprt>(ref.get());
+    const auto current_index = numeric_cast_v<std::size_t>(with_expr.where());
+    entries.emplace_back(current_index, with_expr.new_value());
+    ref = with_expr.old();
+  }
+
+  INVARIANT(
+    ref.get().id() == ID_array_of, "'with_exprt' should contain 'array_of'");
```

Comment:
I don't agree with this statement in the general case but I can believe that it is true in this specific case.  Any chance of some kind of comment saying why?

Reply:
Yes that's right, this is a precondition of this function rather than an invariant of the program. I've updated this and the comment of the functions.